src/main/java/com/gogym/member/service/KakaoService.java

```diff
@@ -0,0 +1,127 @@
+import lombok.extern.slf4j.Slf4j;
+
+@Service
+@RequiredArgsConstructor
+@Transactional(readOnly = true)
+@Slf4j
+public class KakaoService {
+
+  private final MemberRepository memberRepository;
+  private final KakaoMemberRepository kakaoMemberRepository;
+  private final JwtTokenProvider jwtTokenProvider;
+
+  @Value("${kakao.rest-api-key}")
+  private String kakaoClientId;
+
+  @Value("${kakao.redirect-uri}")
+  private String redirectUri;
+
+  @Transactional
+  public String handleKakaoCallback(String code) {
+    KakaoTokenResponse tokenResponse = getAccessToken(code);
+    KakaoProfileResponse profileResponse = getProfile(tokenResponse.getAccessToken());
+    Member member = findOrCreateMember(profileResponse, tokenResponse);
+    String token = jwtTokenProvider.createToken(member.getEmail(), member.getId(),
+        List.of(member.getRole().name()));
+
+    return token;
+  }
+
+  private KakaoTokenResponse getAccessToken(String code) {
+    RestTemplate restTemplate = new RestTemplate();
+    HttpHeaders headers = new HttpHeaders();
+    headers.set("Content-Type", "application/x-www-form-urlencoded");
+
+    String body =
+        String.format("grant_type=authorization_code&client_id=%s&redirect_uri=%s&code=%s",
+            kakaoClientId, redirectUri, code);
+
+    HttpEntity<String> request = new HttpEntity<>(body, headers);
+    ResponseEntity<KakaoTokenResponse> response = restTemplate.exchange(
+        "https://kauth.kakao.com/oauth/token", HttpMethod.POST, request, KakaoTokenResponse.class);
+    return response.getBody();
+  }
+
+  private KakaoProfileResponse getProfile(String accessToken) {
+    RestTemplate restTemplate = new RestTemplate();
+    HttpHeaders headers = new HttpHeaders();
+    headers.set("Authorization", "Bearer " + accessToken);
+
+    HttpEntity<Void> request = new HttpEntity<>(headers);
+    ResponseEntity<KakaoProfileResponse> response = restTemplate.exchange(
+        "https://kapi.kakao.com/v2/user/me", HttpMethod.GET, request, KakaoProfileResponse.class);
+    return response.getBody();
+  }
+
+  @Transactional
+  private KakaoMember registerKakaoMember(KakaoProfileResponse profileResponse) {
+    Member newMember = Member.builder().email(profileResponse.getKakaoAccount().getEmail())
+        .name(profileResponse.getKakaoAccount().getEmail().split("@")[0])
+        .nickname("고짐이_" + (int) (Math.random() * 10000)).isKakao(true).memberStatus("APPROVED")
+        .build();
+
+    memberRepository.save(newMember);
+
+    KakaoMember kakaoMember = KakaoMember.builder().kakaoId(profileResponse.getId())
+        .member(newMember).uuid(java.util.UUID.randomUUID().toString()).build();
+
+    return kakaoMemberRepository.save(kakaoMember);
+  }
+
+  @Transactional
+  private Member findOrCreateMember(KakaoProfileResponse profile,
+      KakaoTokenResponse tokenResponse) {
+    if (profile.getKakaoAccount() == null
+        || !Boolean.TRUE.equals(profile.getKakaoAccount().getHasEmail())) {
```

Comment:
false를 확인하는 거라면
```suggestion
        || !profile.getKakaoAccount().getHasEmail()) {
```
이렇게 사용하는 게 좋지 않을까요?

Reply:
getHasEmail()가 null인 경우 널 방지 오류가 날 수 있어서 현재 코드로 유지하는 것이 더 좋다고 생각합니다.